### PR TITLE
Bug fix/internal issue #726

### PR DESCRIPTION
**Pull request by Dronplane**
PR adds special processing for _analyzers collection during restore/replication.
For single/AF/MS cases just forces analyzers cache reload (without it imported analyzers will not be usable)
For cluster case analyzers operations are performed by IResearchAnalyzersFeature with proper agency operations.

Fix is verifed by gtest and js tests.

Also includes fix for internal issue #658 - added tests

### CODE REVIEW

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10884/

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10885/

File: arangod/Replication/TailingSyncer.cpp [hunk 1]
```diff
@@ -1048,12 +1055,30 @@ Result TailingSyncer::applyLog(SimpleHttpResult* response, TRI_voc_tick_t firstR
...
       reloadUsers();
       _usersModified = false;
     }
+    if (!_analyzersModified.empty() && 
+         (*_analyzersModified.begin())
+                             ->server()
+                             .hasFeature<iresearch::IResearchAnalyzerFeature>()) {
+      auto& analyzersFeature = (*_analyzersModified.begin())
+                                                   ->server()
+                                                   .getFeature<iresearch::IResearchAnalyzerFeature>();
```
Comment (gnusi) [inline arangod/Replication/TailingSyncer.cpp:1071]: let's make it a bit more readable
```
    if (!_analyzersModified.empty()) {
      TRI_ASSERT(_analyzersModified.begin());
      auto& begin = *_analyzersModified.begin();
      auto& server = begin->server();
      if (server.hasFeature<iresearch::IResearchAnalyzerFeature>()) {
        auto& analyzersFeature = server().getFeature<iresearch::IResearchAnalyzerFeature>();
```

File: arangod/Replication/TailingSyncer.cpp [hunk 2]
```diff
@@ -391,10 +392,9 @@ Result TailingSyncer::processDocument(TRI_replication_operation_e type,
   if (coll == nullptr) {
     return Result(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND);
   }
-
   bool const isSystem = coll->system();
   bool const isUsers = coll->name() == TRI_COL_NAME_USERS; 
-
+  bool const isAnalyzers = coll->name() == StaticStrings::AnalyzersCollection;
```
Comment (gnusi) [inline arangod/Replication/TailingSyncer.cpp:397]: I'd switch to `enum`

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10892/

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 3]
```diff
@@ -1644,6 +1717,227 @@ Result IResearchAnalyzerFeature::emplace(
...
+    auto stringRev = std::to_string(analyzerModificationTrx->buildingRevision());
+    std::string aql("FOR u IN ");
+    aql.append(arangodb::StaticStrings::AnalyzersCollection).append(" UPDATE u.")
+       .append(arangodb::StaticStrings::KeyString).append(" WITH { ")
+       .append(arangodb::StaticStrings::AnalyzersDeletedRevision)
+       .append(": ").append(stringRev).append("} IN ")
+       .append(arangodb::StaticStrings::AnalyzersCollection);
+
+    auto ctx = transaction::StandaloneContext::Create(vocbase);
+    SingleCollectionTransaction trx(ctx, arangodb::StaticStrings::AnalyzersCollection,
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:1764]: I am not sure if there are side-effects here when having two `SingleCollectionTransaction` objects in the same scope.
The reason is that the destructor of the base class `transaction::Methods` does some cleanup, and I am not 100% sure this is safe when having two such objects in scope at the same time.
So for safety I would prefer putting each `SingleCollectionTransaction` object in its own nested scope here. That will definitely be safe.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 4]
```diff
@@ -1644,6 +1717,227 @@ Result IResearchAnalyzerFeature::emplace(
...
+    SCOPED_LOCK(mutex);
+
+    if (transaction) {
+      auto cleanupResult = cleanupAnalyzersCollection(vocbase.name(),
+                                                      transaction->buildingRevision());
+      if (cleanupResult.fail()) {
+        return cleanupResult;
+      }
+    }
+    auto res = loadAnalyzers(vocbase.name());
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:1827]: mmmh, as far as I can see `loadAnalyzers()` also acquires the write-mutex on `_mutex`: https://github.com/arangodb/arangodb/pull/12059/files#diff-2eae74915725dafeff75a70e1882fafcL1981
So the mutex would be acquired recursively. Is that safe to do?

Comment (Dronplane) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp:1827]: It is safe as mutex used supports recursive locking. However I guess it is a good point to move lock lower to allow access to analyzers while  cleanup operation runs (as cleanup don`t touch already cached analyzers it is safe)

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 5]
```diff
@@ -1870,6 +1871,10 @@ Result DatabaseInitialSyncer::handleCollection(VPackSlice const& parameters,
 
     if (masterName == TRI_COL_NAME_USERS) {
       reloadUsers();
+    } else if (masterName == StaticStrings::AnalyzersCollection &&
+               ServerState::instance()->isSingleServer() &&
+               vocbase().server().hasFeature<iresearch::IResearchAnalyzerFeature>()) {
```
[the remark this answers is not in the document]
Comment (Dronplane) [reply arangod/Replication/DatabaseInitialSyncer.cpp:1876]: Maybe this is done for tests as server mock often has only subset of features? I mean this is general pattern almost everywhere hasFeature->getFeature unless there are some "assumptions" feature is present.  I`ll be glad to remove checks if I`m wrong with this.

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10896/

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 6]
```diff
@@ -1644,6 +1716,231 @@ Result IResearchAnalyzerFeature::emplace(
...
+    { // SingleCollectionTransaction scope
+      // now let`s do cleanup
+      SingleCollectionTransaction truncateTrx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                              AccessMode::Type::EXCLUSIVE);
+
+      auto res = truncateTrx.begin();
+
+      if (res.ok()) {
+        OperationOptions options;
+        truncateTrx.truncateAsync(arangodb::StaticStrings::AnalyzersCollection, options).get();
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:1794]: Yet another question: this truncates the `_analyzers` collection, but a few lines above there is an UPDATE query executes on it.
Wouldn't it be easier to only run truncate and get rid of the UPDATE query?
As far as I can see, everything that is updated will be removed by the truncate directly after anyway. Or am I missing something?

Comment (Dronplane) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp:1794]: You are missing analyzerModificationTransaction commit which is agency operation. The idea is to first mark 
analyzers as deleted starting with some revision number. And than commit that number into agency. If agency commit fails, we a save as analyzers still present and delete mark will eventually be cleared by cleanup operations. And only after successful commit we actually remove effectively dead analyzers. Even if removal fails it is ok. Leftovers will be eventually cleaned by cleanup. So why there 2 step operation.

Comment (jsteemann) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp:1794]: Got it, thanks for the explanation!

File: arangod/Replication/DatabaseInitialSyncer.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Replication/DatabaseInitialSyncer.cpp:1876]: As the extra check is safe and does no harm, we can definitely leave it here.

File: arangod/Replication/TailingSyncer.cpp [hunk 7]
```diff
@@ -1048,12 +1055,30 @@ Result TailingSyncer::applyLog(SimpleHttpResult* response, TRI_voc_tick_t firstR
...
     if (_usersModified) {
       // reload users after initial dump
       reloadUsers();
       _usersModified = false;
     }
+    if (!_analyzersModified.empty()) {
+      TRI_ASSERT(*_analyzersModified.begin());
+      auto& begin = *_analyzersModified.begin();
+      auto& server = begin->server();
+      if (server.hasFeature<iresearch::IResearchAnalyzerFeature>()) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Replication/TailingSyncer.cpp:1069]: As the extra check is safe and does no harm, we can definitely leave it here.

**Comment by jsteemann** [comment]
I saw that now there are test errors, probably due to moving the SingleCollectionTransactions into scopes of their own. While I still think this is fine, I think it is also required to move the transaction::Context object into the scope(s) as well, so that each transaction gets its own Context instance.

**Comment by Dronplane** [comment]
Ok, thx for hint. Will fix on monday.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10898/

**Comment by jsteemann** [comment]
Tests blue

**Pull request merged**